crowbar_framework/app/models/service_object.rb

```diff
@@ -1713,4 +1714,34 @@ def active_update(proposal, inst, in_queue, bootstrap = false)
       [400, e2.message]
     end
   end
+
+  def skip_unready_nodes(bc, inst, elements)
+    skip_unready_nodes_config = Rails.application.config.experimental["skip_unready_nodes"]
+    # return if the feature is disabled
+    return elements unless skip_unready_nodes_config["enabled"]
+    logger.warn("Experimental skip_unready_nodes feature enabled")
+    logger.debug("skip_unready_nodes: enter for #{bc}:#{inst}")
+    cleaned_elements = elements.deep_dup
+    skip_unready_nodes_config["roles"].each do |role|
+      next unless elements.key?(role)
+      elements[role].each do |n|
+        # TODO: deal with clusters...how?????
+        node = Node.find_by_name(n)
+        next if node.nil?
+        # skip if nodes are on ready or crowbar_upgrade state, we dont need to do anything
+        next if ["ready", "crowbar_upgrade"].include?(node.crowbar["state"])
+        logger.warn(
+          "Node #{n} is being removed from the deployment for #{bc}:#{inst} with role #{role}"
+        )
+        logger.warn(
+          "Node #{n} state is: #{node.crowbar["state"]}"
+        )
+        # TODO: raise an error if this is the only node on the role and its being skipped?
```

Comment:
Why? It will throw errors when you enable skipping for "role1" but you didn't want to deploy this role at all. Note that nova-compute-* roles are mostly exclusive so you won't be able to enable it for more than one without getting errors.

Reply:
I was thinking that if we have a role with just one node in it, and there is some validation on crowbar-framework to make sure that there is at least 1 node for that role before deploying the barclamp, if we just remove that, then the deployment will fail, probably with an obscure error as it expects at least one node there, so in this case we may be should fail before continuing the execution?

I dont think this affects nova-compute-kvm but maybe it affects dns-client (the other role that this feature is targeted to)

Reply:
Do you have a link to some code which does such validation?

Reply:
https://github.com/crowbar/crowbar-openstack/search?utf8=%E2%9C%93&q=validate_at_least_n_for_role&type=

Reply:
To be honest, I don't think it's an issue and that's related to my earlier comment: if we only skip unready nodes that were already in the proposal before, then there will still be a node with the role -- even if chef-client is not run with the apply.